soda-core/src/soda_core/common/soda_cloud.py

```diff
@@ -1196,3 +1198,25 @@ def _append_exception_to_cloud_log_dicts(cloud_log_dicts: list[dict], exception:
     exc_cloud_log_dict["index"] = len(cloud_log_dicts)
     cloud_log_dicts.append(exc_cloud_log_dict)
     return cloud_log_dicts
+
+
+class FailedRowsDiagnostic:
```

Comment:
I would use pydantic to skip having to create the constructors, especially the duplicate ones in the subclasses. Actually - all the 4 constructors are identical, there is no need to repeat them in the subclasses